File: LeetCode/Kai/279PerfectSquares.java
public class Solution {
    public int numSquares(int n) {
        int[] dp = new int[n+1];
        Arrays.fill(dp, Integer.MAX_VALUE);
        for(int i=1; i*i<=n; i++) dp[i*i] = 1;   //base
        
        for(int i=1; i<=n; i++) {
            for(int j=1; i+j*j<=n; j++) {
                dp[i+j*j] = Math.min(dp[i+j*j], dp[i]+1);
            }
        }
        
        return dp[n];
        
    }
    
}

File: LeetCode/Liang/279PerfectSquares.cpp
class Solution {
public:
    int numSquares(int n) {
        vector<int> DP(n + 1, INT_MAX);
        DP.at(0) = 0;
        for (int i = 1; i <= n; ++i) {
            for (int j = 1; j * j <= i; ++j) {
                DP.at(i) = min(DP.at(i), DP.at(i - j * j) + 1);
            }
        }
        return DP.back();
    }
};
